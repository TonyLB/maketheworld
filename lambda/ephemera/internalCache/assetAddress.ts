import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { CacheConstructor } from './baseClasses'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace/ts/readOnly'

export type AssetAddressItem = {
    EphemeraId: EphemeraAssetId;
    address: AssetWorkspaceAddress
}

export class CacheAssetAddressData {
    AssetAddressById: Record<EphemeraAssetId, AssetAddressItem> = {};
    clear() {
        this.AssetAddressById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<AssetAddressItem | undefined> {
        return this.AssetAddressById[assetId]
    }
    set(assetItem: AssetAddressItem): void {
        this.AssetAddressById[assetItem.EphemeraId] = assetItem
    }
}

export const CacheAssetAddress = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetAddress extends Base {
        AssetAddress: CacheAssetAddressData = new CacheAssetAddressData()

        override clear() {
            this.AssetAddress.clear()
            super.clear()
        }
    }
}

export default CacheAssetAddress
