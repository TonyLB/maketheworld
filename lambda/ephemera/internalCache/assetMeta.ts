import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export type AssetMetaItem = {
    EphemeraId: EphemeraAssetId;
    found: boolean;
}

export class CacheAssetMetaData {
    AssetMetaById: Record<EphemeraAssetId, AssetMetaItem> = {};
    clear() {
        this.AssetMetaById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<Omit<AssetMetaItem, 'found'> | undefined> {
        if (!(this.AssetMetaById[assetId])) {
            const assetData = await ephemeraDB.getItem<AssetMetaItem>({
                    Key: {
                        EphemeraId: assetId,
                        DataCategory: 'Meta::Asset'
                    },
                    ProjectionFields: ['EphemeraId'],
                })
            if (assetData) {
                this.AssetMetaById[assetId] = {
                    ...assetData,
                    found: true
                }
            }
            else {
                this.AssetMetaById[assetId] = {
                    EphemeraId: assetId,
                    found: false
                }
            }
        }
        if (this.AssetMetaById[assetId]?.found) {
            const { found, ...rest } = this.AssetMetaById[assetId]
            return rest
        }
        else {
            return undefined
        }
    }
    set(assetItem: Omit<AssetMetaItem, 'found'>): void {
        this.AssetMetaById[assetItem.EphemeraId] = {
            ...assetItem,
            found: true
        }
    }
}

export const CacheAssetMeta = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetMeta extends Base {
        AssetMeta: CacheAssetMetaData = new CacheAssetMetaData()

        override clear() {
            this.AssetMeta.clear()
            super.clear()
        }
    }
}

export default CacheAssetMeta
