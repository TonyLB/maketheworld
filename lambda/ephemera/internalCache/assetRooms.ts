import { EphemeraAssetId, EphemeraRoomId } from '@tonylb/mtw-interfaces/dist/baseClasses'
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export type AssetRoomsItem = {
    EphemeraId: EphemeraAssetId;
    rooms: EphemeraRoomId[];
    found: boolean;
}

export class CacheAssetRoomsData {
    AssetRoomsById: Record<EphemeraAssetId, AssetRoomsItem> = {};
    clear() {
        this.AssetRoomsById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<Omit<AssetRoomsItem, 'found'> | undefined> {
        if (!(this.AssetRoomsById[assetId])) {
            const assetRooms = await ephemeraDB.query<{ EphemeraId: EphemeraRoomId }[]>({
                IndexName: 'DataCategoryIndex',
                DataCategory: assetId,
                KeyConditionExpression: 'begins_with(EphemeraId, :roomPrefix)',
                ExpressionAttributeValues: {
                    ':roomPrefix': 'ROOM#'
                },        
            })
            if (assetRooms) {
                this.AssetRoomsById[assetId] = {
                    EphemeraId: assetId,
                    rooms: assetRooms.map(({ EphemeraId }) => (EphemeraId)),
                    found: true
                }
            }
            else {
                this.AssetRoomsById[assetId] = {
                    EphemeraId: assetId,
                    rooms: [],
                    found: false
                }
            }
        }
        if (this.AssetRoomsById[assetId]?.found) {
            const { found, ...rest } = this.AssetRoomsById[assetId]
            return rest
        }
        else {
            return undefined
        }
    }
    set(assetItem: Omit<AssetRoomsItem, 'found'>): void {
        this.AssetRoomsById[assetItem.EphemeraId] = {
            ...assetItem,
            found: true
        }
    }
}

export const CacheAssetRooms = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetRooms extends Base {
        AssetRooms: CacheAssetRoomsData = new CacheAssetRoomsData()

        override clear() {
            this.AssetRooms.clear()
            super.clear()
        }
    }
}

export default CacheAssetRooms
