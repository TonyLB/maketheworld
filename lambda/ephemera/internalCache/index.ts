import { CacheConstructor, CacheBase } from './baseClasses'

import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'
import { delayPromise } from '@tonylb/mtw-utilities/dist/dynamoDB/delayPromise'
import CacheRoomCharacterLists from './roomCharacterLists';
import CacheCharacterMeta from './characterMeta';
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB';
import CacheCharacterConnections from './characterConnections';
import AssetState from './assetState';
import DependencyGraph from './dependencyGraph';
import ComponentMeta from './componentMeta';
import ComponentRender from './componentRender';

type CacheGlobalKeys = 'ConnectionId' | 'RequestId' | 'player' | 'assets' | 'connections'
class CacheGlobalData {
    ConnectionId?: string;
    RequestId?: string;
    player?: string;
    assets?: string[];
    connections?: string[];
    get(key: 'ConnectionId' | 'RequestId' | 'player'): Promise<string | undefined>
    get(key: 'assets' | 'connections'): Promise<string[] | undefined>
    get(key: CacheGlobalKeys): Promise<string | string[] | undefined>
    async get(key: CacheGlobalKeys) {
        switch(key) {
            case 'player':
                if (!this.player && this.ConnectionId) {
                    //
                    // TODO: Replace repeated attempts with exponential backoff by
                    // refactoring connectionDB.getItem to allow a consistent argument
                    // that can actviate strongly-consistent reads
                    //
                    let attempts = 0
                    let exponentialBackoff = 50
                    while(attempts < 5) {
                        const { player = '' } = await connectionDB.getItem<{ player: string }>({
                            ConnectionId: `CONNECTION#${this.ConnectionId}`,
                            DataCategory: 'Meta::Connection',
                            ProjectionFields: ['player']
                        }) || {}
                        if (player) {
                            this.player = player
                            return player
                        }
                        attempts += 1
                        await delayPromise(exponentialBackoff)
                        exponentialBackoff = exponentialBackoff * 2
                    }
                    console.log(`Exponential backoff on player caching failed after five attempts (${this.ConnectionId})`)
                }
                return this.player
            case 'assets':
                if (typeof this.assets === 'undefined') {
                    const { assets = [] } = (await ephemeraDB.getItem<{ assets: string[] }>({
                        EphemeraId: 'Global',
                        DataCategory: 'Assets',
                        ProjectionFields: ['assets']
                    })) || {}
                    this.assets = assets
                }
                return this.assets
            case 'connections':
                if (typeof this.connections === 'undefined') {
                    const { connections = {} } = (await connectionDB.getItem<{ connections: Record<string, string> }>({
                        ConnectionId: 'Global',
                        DataCategory: 'Connections',
                        ProjectionFields: ['connections']
                    })) || {}
                    this.connections = Object.keys(connections)
                }
                return this.connections
            default:
                return this[key]
        }
    }

    clear() {
        this.ConnectionId = undefined
        this.RequestId = undefined
        this.player = undefined
        this.assets = undefined
    }

    set(props: { key: 'ConnectionId' | 'RequestId', value: string; }): void {
        this[props.key] = props.value
    }
}

export const CacheGlobal = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Global: CacheGlobalData = new CacheGlobalData()

        override clear() {
            this.Global.clear()
            super.clear()
        }
    }
}

const InternalCache = ComponentRender(ComponentMeta(AssetState(DependencyGraph(CacheCharacterConnections(CacheCharacterMeta(CacheRoomCharacterLists(CacheGlobal(CacheBase))))))))
export const internalCache = new InternalCache()
export default internalCache
