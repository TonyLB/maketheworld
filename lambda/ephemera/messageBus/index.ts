import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'

import {
    MessageBus,
    isPublishMessage,
    isEphemeraUpdate,
    isDisconnectMessage,
    isConnectMessage,
    isWhoAmIMessage,
    isSyncRequest,
    isSyncResponse,
    isRegisterCharacterMessage,
    isFetchPlayerEphemera,
    isImportDefaults,
    isFetchImportDefaults,
    isPerception,
    isMoveCharacter,
    isDecacheAsset,
    isCacheAssetMessage,
    isPlayerUpdateMessage,
    isRoomUpdateMessage,
    isDescentUpdateMessage,
    isAncestryUpdateMessage,
    isDependencyCascadeMessage,
    isExecuteActionMessage,
    isMapSubscription,
    isMapUpdateMessage
} from "./baseClasses"

import publishMessage from '../publishMessage'
import ephemeraUpdate from '../ephemeraUpdate'
import disconnectMessage from '../disconnectMessage'
import connectMessage from '../connectMessage'
import whoAmIMessage from '../whoAmI'
import { syncRequest, syncResponse } from '../syncHandler'
import registerCharacter from '../registerCharacter'
import { fetchPlayerEphemera } from '../fetchEphemera'
import perceptionMessage from '../perception'
import moveCharacter from '../moveCharacter'
import decacheAssetMessage from '../decacheMessage'
import { cacheAssetMessage } from '../cacheAsset'
import playerUpdateMessage from '../playerUpdate'
import roomUpdateMessage from '../roomUpdate'
import dependentUpdateMessage from '../dependentMessages/dependentUpdate'
import dependencyCascadeMessage from '../dependentMessages/dependencyCascade'
import executeActionMessage from '../executeAction'
import mapSubscriptionMessage from '../mapSubscription'
import mapUpdateMessage from '../mapUpdate'

export const messageBus = new MessageBus()
messageBus.subscribe({
    tag: 'PublishMessage',
    priority: 15,
    filter: isPublishMessage,
    callback: publishMessage
})
messageBus.subscribe({
    tag: 'EphemeraUpdate',
    priority: 10,
    filter: isEphemeraUpdate,
    callback: ephemeraUpdate
})
messageBus.subscribe({
    tag: 'Disconnect',
    priority: 1,
    filter: isDisconnectMessage,
    callback: disconnectMessage
})
messageBus.subscribe({
    tag: 'Connect',
    priority: 1,
    filter: isConnectMessage,
    callback: connectMessage
})
messageBus.subscribe({
    tag: 'WhoAmI',
    priority: 1,
    filter: isWhoAmIMessage,
    callback: whoAmIMessage
})
messageBus.subscribe({
    tag: 'SyncResponse',
    priority: 2,
    filter: isSyncResponse,
    callback: syncResponse
})
messageBus.subscribe({
    tag: 'SyncRequest',
    priority: 3,
    filter: isSyncRequest,
    callback: syncRequest
})
messageBus.subscribe({
    tag: 'RegisterCharacter',
    priority: 1,
    filter: isRegisterCharacterMessage,
    callback: registerCharacter
})
messageBus.subscribe({
    tag: 'FetchPlayerEphemera',
    priority: 2,
    filter: isFetchPlayerEphemera,
    callback: fetchPlayerEphemera
})
messageBus.subscribe({
    tag: 'MapSubscription',
    priority: 2,
    filter: isMapSubscription,
    callback: mapSubscriptionMessage
})
messageBus.subscribe({
    tag: 'Perception',
    priority: 10,
    filter: isPerception,
    callback: perceptionMessage
})
messageBus.subscribe({
    tag: 'MoveCharacter',
    priority: 4,
    filter: isMoveCharacter,
    callback: moveCharacter
})
messageBus.subscribe({
    tag: 'DecacheAsset',
    priority: 1,
    filter: isDecacheAsset,
    callback: decacheAssetMessage
})
messageBus.subscribe({
    tag: 'CacheAsset',
    priority: 1,
    filter: isCacheAssetMessage,
    callback: cacheAssetMessage
})
messageBus.subscribe({
    tag: 'UpdatePlayer',
    priority: 5,
    filter: isPlayerUpdateMessage,
    callback: playerUpdateMessage
})
messageBus.subscribe({
    tag: 'RoomUpdate',
    priority: 6,
    filter: isRoomUpdateMessage,
    callback: roomUpdateMessage
})
messageBus.subscribe({
    tag: 'MapUpdate',
    priority: 15,
    filter: isMapUpdateMessage,
    callback: mapUpdateMessage
})
messageBus.subscribe({
    tag: 'AncestryUpdate',
    priority: 3,
    filter: isAncestryUpdateMessage,
    callback: dependentUpdateMessage('Ancestry')
})
messageBus.subscribe({
    tag: 'DescentUpdate',
    priority: 4,
    filter: isDescentUpdateMessage,
    callback: dependentUpdateMessage('Descent')
})

messageBus.subscribe({
    tag: 'DependencyCascade',
    priority: 5,
    filter: isDependencyCascadeMessage,
    callback: dependencyCascadeMessage
})

messageBus.subscribe({
    tag: 'ExecuteAction',
    priority: 7,
    filter: isExecuteActionMessage,
    callback: executeActionMessage
})

export default messageBus