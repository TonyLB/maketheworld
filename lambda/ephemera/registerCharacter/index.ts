import { RegisterCharacterMessage, MessageBus } from "../messageBus/baseClasses"
import messageBus from "../messageBus"
import { ephemeraDB, connectionDB, exponentialBackoffWrapper, multiTableTransactWrite } from "@tonylb/mtw-utilities/dist/dynamoDB"

import internalCache from '../internalCache'
import { unique } from "@tonylb/mtw-utilities/dist/lists"
import { marshall } from "@aws-sdk/util-dynamodb"
import { splitType } from "@tonylb/mtw-utilities/dist/types"
import { RoomCharacterListItem } from "../internalCache/baseClasses"
import { EphemeraRoomId, isEphemeraCharacterId } from "@tonylb/mtw-interfaces/dist/baseClasses"

export const registerCharacter = async ({ payloads }: { payloads: RegisterCharacterMessage[], messageBus: MessageBus }): Promise<void> => {

    const connectionId = await internalCache.Global.get('ConnectionId')

    if (connectionId) {
        const RequestId = await internalCache.Global.get('RequestId')
        const handleOneRegistry = async (payload: RegisterCharacterMessage): Promise<void> => {
            const { characterId: CharacterId } = payload
            if (!(isEphemeraCharacterId(CharacterId))) {
                return
            }
            await exponentialBackoffWrapper(async () => {
                const [characterFetch, currentConnections] = await Promise.all([
                    internalCache.CharacterMeta.get(CharacterId),
                    internalCache.CharacterConnections.get(CharacterId)
                ])
                if (!characterFetch) {
                    return
                }
                const { Name = '', HomeId = '', RoomId = '', fileURL, Color } = characterFetch
                const RoomEphemeraId: EphemeraRoomId = `ROOM#${RoomId || HomeId || 'VORTEX'}`
                const activeCharacters = await internalCache.RoomCharacterList.get(splitType(RoomEphemeraId)[1])
                const newConnections = unique(currentConnections || [], [connectionId]) as string[]
                const metaCharacterUpdate = (typeof currentConnections !== 'undefined')
                    ? {
                        TableName: 'Connections',
                        Key: marshall({
                            ConnectionId: CharacterId,
                            DataCategory: 'Meta::Character'
                        }),
                        ExpressionAttributeValues: marshall({
                            ':oldConnections': currentConnections,
                            ':newConnections': newConnections
                        }),
                        UpdateExpression: 'SET connections = :newConnections',
                        ConditionExpression: 'connections = :oldConnections'
                    }
                    : {
                        TableName: 'Connections',
                        Key: marshall({
                            ConnectionId: CharacterId,
                            DataCategory: 'Meta::Character'
                        }),
                        ExpressionAttributeValues: marshall({
                            ':newConnections': newConnections
                        }),
                        UpdateExpression: 'SET connections = :newConnections',
                        ConditionExpression: 'attribute_not_exists(connections)'
                    }
                const characterRoomUpdate = RoomId
                    ? []
                    : [{
                        Update: {
                            TableName: 'Ephemera',
                            Key: marshall({
                                EphemeraId: CharacterId,
                                DataCategory: 'Meta::Character',
                            }),
                            UpdateExpression: 'SET RoomId = :roomId',
                            ExpressionAttributeValues: marshall({
                                ':roomId': HomeId || 'VORTEX',
                            }),
                            ConditionExpression: 'attribute_not_exists(RoomId)'
                        }
                    }]
                const newActiveCharacters: RoomCharacterListItem[] = [
                    ...(activeCharacters || []).filter((character) => (character.EphemeraId !== CharacterId)),
                    {
                        EphemeraId: CharacterId,
                        Name,
                        fileURL,
                        Color,
                        ConnectionIds: newConnections
                    }
                ]
                const activeCharactersUpdate = {
                    Update: {
                        TableName: 'Ephemera',
                        Key: marshall({
                            EphemeraId: RoomEphemeraId,
                            DataCategory: 'Meta::Room'
                        }),
                        UpdateExpression: 'SET activeCharacters = :newActiveCharacters',
                        ExpressionAttributeValues: marshall({
                            ':oldActiveCharacters': activeCharacters,
                            ':newActiveCharacters': newActiveCharacters
                        }, { removeUndefinedValues: true}),
                        ConditionExpression: 'activeCharacters = :oldActiveCharacters'
                    }
                }
                await multiTableTransactWrite([{
                    Update: metaCharacterUpdate
                },
                {
                    Put: {
                        TableName: 'Connections',
                        Item: marshall({
                            ConnectionId: `CONNECTION#${connectionId}`,
                            DataCategory: CharacterId
                        })
                    }
                },
                ...characterRoomUpdate,
                activeCharactersUpdate
                ])
                if ((currentConnections || []).length === 0) {
                    messageBus.send({
                        type: 'EphemeraUpdate',
                        global: true,
                        updates: [{
                            type: 'CharacterInPlay',
                            CharacterId,
                            Name: Name || '',
                            Connected: true,
                            RoomId,
                            fileURL: fileURL || '',
                            Color: Color || 'grey'
                        }]        
                    })
                    messageBus.send({
                        type: 'PublishMessage',
                        targets: [{ roomId: RoomEphemeraId }, { excludeCharacterId: CharacterId }],
                        displayProtocol: 'WorldMessage',
                        message: [{
                            tag: 'String',
                            value: `${Name || 'Someone'} has connected.`
                        }]
                    })
                    messageBus.send({
                        type: 'RoomUpdate',
                        roomId: RoomId
                    })
                }
                internalCache.RoomCharacterList.set({ key: RoomId, value: newActiveCharacters })

            }, { retryErrors: ['TransactionCanceledException']})
    
        }

        await Promise.all(payloads.map(handleOneRegistry))

        if (payloads.length > 0) {
            messageBus.send({
                type: 'ReturnValue',
                body: {
                    messageType: 'Registration',
                    CharacterId: payloads[0].characterId,
                    RequestId
                }
            })
        }
    }

}

export default registerCharacter
