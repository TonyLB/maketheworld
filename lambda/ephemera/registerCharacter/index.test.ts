jest.mock('@tonylb/mtw-utilities/dist/dynamoDB/index')
import {
    ephemeraDB,
    connectionDB,
    multiTableTransactWrite
} from '@tonylb/mtw-utilities/dist/dynamoDB/index'

jest.mock('../messageBus')
import messageBus from '../messageBus'

jest.mock('../internalCache')
import internalCache from '../internalCache'

const ephemeraDBMock = ephemeraDB as jest.Mocked<typeof ephemeraDB>
const connectionDBMock = connectionDB as jest.Mocked<typeof connectionDB>
const multiTableTransactWriteMock = multiTableTransactWrite as jest.Mock
const messageBusMock = messageBus as jest.Mocked<typeof messageBus>
const internalCacheMock = jest.mocked(internalCache, true)

import registerCharacter from '.'

describe("registerCharacter", () => {
    const transactWriteMockImplementation = (props: Record<string, any>) => async (items) => {
        items.forEach((item) => {
            if ('Update' in item && item.Update.successCallback) {
                item.Update.successCallback(props)
            }
        })
    }

    beforeEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
        internalCacheMock.Global.get.mockResolvedValueOnce('TestConnection').mockResolvedValueOnce('Request123')
    })

    it("should update correctly on first connection", async () => {
        internalCacheMock.CharacterMeta.get.mockResolvedValueOnce({
            EphemeraId: 'CHARACTER#ABC',
            Name: 'Tess',
            RoomId: 'ROOM#TestABC',
            Color: 'purple',
            HomeId: 'ROOM#VORTEX',
            assets: [],
            Pronouns: { subject: 'they', object: 'them', possessive: 'their', adjective: 'theirs', reflexive: 'themself' }
        })
        internalCacheMock.RoomCharacterList.get.mockResolvedValueOnce([{
            EphemeraId: 'CHARACTER#BCD',
            Name: 'TestToo',
            ConnectionIds: ['QRS']
        }])
        internalCacheMock.CharacterConnections.get.mockResolvedValue([])
        connectionDBMock.transactWrite.mockImplementation(transactWriteMockImplementation({ connections: ['TestConnection']}))
        ephemeraDBMock.transactWrite.mockImplementation(transactWriteMockImplementation({
            EphemeraId: 'ROOM#TestABC',
            activeCharacters: [
                {
                    EphemeraId: 'CHARACTER#BCD',
                    Name: 'TestToo',
                    ConnectionIds: ['QRS']
                },
                {
                    EphemeraId: 'CHARACTER#ABC',
                    Name: 'Tess',
                    Color: 'purple',
                    ConnectionIds: ['TestConnection']
                }
            ]
        }))
        await registerCharacter({
            payloads: [{ type: 'RegisterCharacter', characterId: 'CHARACTER#ABC' }],
            messageBus
        })
        expect(connectionDBMock.transactWrite).toHaveBeenCalledWith([
            { Put: { ConnectionId: 'CONNECTION#TestConnection', DataCategory: 'CHARACTER#ABC' } },
            {
                Update: {
                    Key: {
                        ConnectionId: 'CHARACTER#ABC',
                        DataCategory: 'Meta::Character'
                    },
                    updateKeys: ['connections'],
                    updateReducer: expect.any(Function),
                    successCallback: expect.any(Function)
                }
            }
        ])
        expect(ephemeraDBMock.transactWrite).toHaveBeenCalledWith([
            {
                Update: {
                    Key: {
                        EphemeraId: 'ROOM#TestABC',
                        DataCategory: 'Meta::Room'
                    },
                    updateKeys: ['activeCharacters'],
                    updateReducer: expect.any(Function),
                    successCallback: expect.any(Function)
                }
            },
            {
                Update: {
                    Key: {
                        EphemeraId: 'CHARACTER#ABC',
                        DataCategory: 'Meta::Character'
                    },
                    updateKeys: ['RoomId', 'HomeId'],
                    updateReducer: expect.any(Function)
                }
            }
        ])
        expect(messageBusMock.send).toHaveBeenCalledWith({
            type: 'ReturnValue',
            body: {
                messageType: "Registration",
                CharacterId: 'CHARACTER#ABC',
                RequestId: 'Request123'
            }
        })
        expect(messageBusMock.send).toHaveBeenCalledWith({
            type: 'EphemeraUpdate',
            updates: [{
                type: 'CharacterInPlay',
                CharacterId: 'CHARACTER#ABC',
                Connected: true,
                Name: 'Tess',
                RoomId: 'ROOM#TestABC',
                fileURL: '',
                Color: 'purple',
                targets: ['GLOBAL', 'CONNECTION#TestConnection']
            }]
        })
        expect(messageBusMock.send).toHaveBeenCalledWith({
            type: 'PublishMessage',
            targets: ['ROOM#TestABC', '!CHARACTER#ABC'],
            displayProtocol: 'WorldMessage',
            message: [{
                tag: 'String',
                value: `Tess has connected.`
            }]
        })
        expect(messageBusMock.send).toHaveBeenCalledWith({
            type: 'RoomUpdate',
            roomId: 'ROOM#TestABC'
        })
        expect(internalCacheMock.RoomCharacterList.set).toHaveBeenCalledWith({
            key: 'ROOM#TestABC',
            value: [
                {
                    EphemeraId: 'CHARACTER#BCD',
                    Name: 'TestToo',
                    ConnectionIds: ['QRS']
                },
                {
                    EphemeraId: 'CHARACTER#ABC',
                    Name: 'Tess',
                    Color: 'purple',
                    ConnectionIds: ['TestConnection']
                }
            ]
        })
    })

    it("should update correctly on subsequent connections", async () => {
        internalCacheMock.CharacterMeta.get.mockResolvedValueOnce({
            EphemeraId: 'CHARACTER#ABC',
            Name: 'Tess',
            RoomId: 'ROOM#TestABC',
            Color: 'purple',
            HomeId: 'ROOM#VORTEX',
            assets: [],
            Pronouns: { subject: 'they', object: 'them', possessive: 'their', adjective: 'theirs', reflexive: 'themself' }
        })
        internalCacheMock.RoomCharacterList.get.mockResolvedValueOnce([{
                EphemeraId: 'CHARACTER#ABC',
                Name: 'Tess',
                ConnectionIds: ['previous']
            }])
        internalCacheMock.CharacterConnections.get.mockResolvedValue(['previous'])
        connectionDBMock.transactWrite.mockImplementation(transactWriteMockImplementation({ connections: ['previous', 'TestConnection']}))
        ephemeraDBMock.transactWrite.mockImplementation(transactWriteMockImplementation({
            EphemeraId: 'ROOM#TestABC',
            activeCharacters: [
                {
                    EphemeraId: 'CHARACTER#ABC',
                    Name: 'Tess',
                    Color: 'purple',
                    ConnectionIds: ['previous', 'TestConnection']
                }
            ]
        }))
        await registerCharacter({ payloads: [{ type: 'RegisterCharacter', characterId: 'CHARACTER#ABC' }], messageBus })
        expect(connectionDBMock.transactWrite).toHaveBeenCalledWith([
            { Put: { ConnectionId: 'CONNECTION#TestConnection', DataCategory: 'CHARACTER#ABC' } },
            {
                Update: {
                    Key: {
                        ConnectionId: 'CHARACTER#ABC',
                        DataCategory: 'Meta::Character'
                    },
                    updateKeys: ['connections'],
                    updateReducer: expect.any(Function),
                    successCallback: expect.any(Function)
                }
            }
        ])
        expect(ephemeraDBMock.transactWrite).toHaveBeenCalledWith([
            {
                Update: {
                    Key: {
                        EphemeraId: 'ROOM#TestABC',
                        DataCategory: 'Meta::Room'
                    },
                    updateKeys: ['activeCharacters'],
                    updateReducer: expect.any(Function),
                    successCallback: expect.any(Function)
                }
            },
            {
                Update: {
                    Key: {
                        EphemeraId: 'CHARACTER#ABC',
                        DataCategory: 'Meta::Character'
                    },
                    updateKeys: ['RoomId', 'HomeId'],
                    updateReducer: expect.any(Function)
                }
            }
        ])
        expect(messageBusMock.send).toHaveBeenCalledTimes(2)
        expect(messageBusMock.send).toHaveBeenCalledWith({
            type: 'ReturnValue',
            body: {
                messageType: "Registration",
                CharacterId: 'CHARACTER#ABC',
                RequestId: 'Request123'
            }
        })
        expect(messageBusMock.send).toHaveBeenCalledWith({
            type: 'Perception',
            characterId: 'CHARACTER#ABC',
            ephemeraId: 'ROOM#TestABC',
            header: true
        })
        expect(internalCacheMock.RoomCharacterList.set).toHaveBeenCalledWith({
            key: 'ROOM#TestABC',
            value: [{
                EphemeraId: 'CHARACTER#ABC',
                Name: 'Tess',
                Color: 'purple',
                ConnectionIds: ['previous', 'TestConnection']
            }]
        })
    })

})