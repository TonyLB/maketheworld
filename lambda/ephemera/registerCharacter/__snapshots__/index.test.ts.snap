// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerCharacter should update correctly on first connection 1`] = `
Array [
  Object {
    "Update": Object {
      "ConditionExpression": "connections = :oldConnections",
      "ExpressionAttributeValues": Object {
        ":newConnections": Object {
          "L": Array [
            Object {
              "S": "TestConnection",
            },
          ],
        },
        ":oldConnections": Object {
          "L": Array [],
        },
      },
      "Key": Object {
        "ConnectionId": Object {
          "S": "CHARACTER#ABC",
        },
        "DataCategory": Object {
          "S": "Meta::Character",
        },
      },
      "TableName": "Connections",
      "UpdateExpression": "SET connections = :newConnections",
    },
  },
  Object {
    "Put": Object {
      "Item": Object {
        "ConnectionId": Object {
          "S": "CONNECTION#TestConnection",
        },
        "DataCategory": Object {
          "S": "CHARACTER#ABC",
        },
      },
      "TableName": "Connections",
    },
  },
  Object {
    "Update": Object {
      "ConditionExpression": "activeCharacters = :oldActiveCharacters",
      "ExpressionAttributeValues": Object {
        ":newActiveCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "QRS",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#BCD",
                },
                "Name": Object {
                  "S": "TestToo",
                },
              },
            },
            Object {
              "M": Object {
                "Color": Object {
                  "S": "purple",
                },
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "TestConnection",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#ABC",
                },
                "Name": Object {
                  "S": "Tess",
                },
              },
            },
          ],
        },
        ":oldActiveCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "QRS",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#BCD",
                },
                "Name": Object {
                  "S": "TestToo",
                },
              },
            },
          ],
        },
      },
      "Key": Object {
        "DataCategory": Object {
          "S": "Meta::Room",
        },
        "EphemeraId": Object {
          "S": "ROOM#TestABC",
        },
      },
      "TableName": "Ephemera",
      "UpdateExpression": "SET activeCharacters = :newActiveCharacters",
    },
  },
]
`;

exports[`registerCharacter should update correctly on subsequent connections 1`] = `
Array [
  Object {
    "Update": Object {
      "ConditionExpression": "connections = :oldConnections",
      "ExpressionAttributeValues": Object {
        ":newConnections": Object {
          "L": Array [
            Object {
              "S": "previous",
            },
            Object {
              "S": "TestConnection",
            },
          ],
        },
        ":oldConnections": Object {
          "L": Array [
            Object {
              "S": "previous",
            },
          ],
        },
      },
      "Key": Object {
        "ConnectionId": Object {
          "S": "CHARACTER#ABC",
        },
        "DataCategory": Object {
          "S": "Meta::Character",
        },
      },
      "TableName": "Connections",
      "UpdateExpression": "SET connections = :newConnections",
    },
  },
  Object {
    "Put": Object {
      "Item": Object {
        "ConnectionId": Object {
          "S": "CONNECTION#TestConnection",
        },
        "DataCategory": Object {
          "S": "CHARACTER#ABC",
        },
      },
      "TableName": "Connections",
    },
  },
  Object {
    "Update": Object {
      "ConditionExpression": "activeCharacters = :oldActiveCharacters",
      "ExpressionAttributeValues": Object {
        ":newActiveCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "Color": Object {
                  "S": "purple",
                },
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "previous",
                    },
                    Object {
                      "S": "TestConnection",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#ABC",
                },
                "Name": Object {
                  "S": "Tess",
                },
              },
            },
          ],
        },
        ":oldActiveCharacters": Object {
          "L": Array [
            Object {
              "M": Object {
                "ConnectionIds": Object {
                  "L": Array [
                    Object {
                      "S": "previous",
                    },
                  ],
                },
                "EphemeraId": Object {
                  "S": "CHARACTER#ABC",
                },
                "Name": Object {
                  "S": "Tess",
                },
              },
            },
          ],
        },
      },
      "Key": Object {
        "DataCategory": Object {
          "S": "Meta::Room",
        },
        "EphemeraId": Object {
          "S": "ROOM#TestABC",
        },
      },
      "TableName": "Ephemera",
      "UpdateExpression": "SET activeCharacters = :newActiveCharacters",
    },
  },
]
`;
